### PR TITLE
#669 default configuration not working when patients change

### LINKED ISSUE #669: Default configuration not working when patients change
When patients are referenced in the default config,
And then some patients are removed, the computations fails. 
Make this more robust.

### PR DESCRIPTION
**Pull request by krllstdn**
Closes #669

### CODE REVIEW

**Comment by krllstdn** [comment]
This PR is still a work in progress and is paused for some time. Some tests that modulate the problem were made, though I'm not sure how well they do their work.

**Review by kubantjan — CHANGES_REQUESTED** [review]
ten problem nastava nejen kdyz se pacient smaze ale i kdyz se jen pacient stane nevalidnim. 

je to potreba jeste dotahnout. Prijde mi ze mozna bude lepsi, nez to handlovat na urovni mist, kde se to muze zmenit ten pacient (delete donor nebo modify) to delat na miste, kde se config aplikuje a proste nevalidni veci z configu ignorovat.

File: Makefile [hunk 1]
```diff
@@ -159,3 +159,5 @@ create-demo-event-json:
 		echo $${out}: ;\
 		cat $${out} ; \
 	done; \
+
+rerun: setup-non-empty-db run
```
Comment (kristinagalik) [inline Makefile:163]: missing endline

File: txmatching/configuration/configuration.py [hunk 2]
```diff
@@ -1,9 +1,9 @@
 from dataclasses import dataclass
 
 from txmatching.configuration.config_parameters import ConfigParameters
-from txmatching.database.services.txm_event_service import \
-    raise_error_if_txm_event_does_not_exist
-from txmatching.database.sql_alchemy_schema import ConfigModel
+from txmatching.data_transfer_objects import txm_event
+from txmatching.database.services.txm_event_service import get_txm_event_complete, raise_error_if_txm_event_does_not_exist
```
Comment (kristinagalik) [inline txmatching/configuration/configuration.py:5]: why import `get_txm_event_complete`? i think its not used anywhere

Comment (kristinagalik) [reply txmatching/configuration/configuration.py:5]: same goes for `RecipientModel`

File: txmatching/solve_service/solve_from_configuration.py [hunk 3]
```diff
@@ -21,6 +21,16 @@ def solve_from_configuration(config_parameters: ConfigParameters, txm_event: Txm
...
 
 def _solve_from_configuration_unsafe(config_parameters: ConfigParameters, txm_event: TxmEvent) -> PairingResult:
+    # check if there are invalid patients in the config
+    for patient_db_id in config_parameters.required_patient_db_ids:
+        if patient_db_id not in txm_event.active_and_valid_recipients_dict:
+            config_parameters.required_patient_db_ids.remove(patient_db_id)
+
+    for pair in config_parameters.manual_donor_recipient_scores:
+        if pair.donor_db_id not in txm_event.active_and_valid_donors_dict \
+            and pair.recipient_db_id not in txm_event.active_and_valid_recipients_dict:
```
Comment (kristinagalik) [inline txmatching/solve_service/solve_from_configuration.py:31]: it should be probably `or` rather than `and`

**Review by kristinagalik — CHANGES_REQUESTED** [review]
few small comments :) otherwise ok

**Review by kristinagalik — DISMISSED** [review]

File: txmatching/database/services/patient_service.py [hunk 4]
```diff
@@ -422,6 +422,18 @@ def delete_donor_recipient_pair(donor_id: int, txm_event_id: int):
         delete_parsing_issues_for_patient(recipient_id=maybe_recipient.db_id, txm_event_id=txm_event_id)
         RecipientModel.query.filter(RecipientModel.id == maybe_recipient.db_id).delete()
 
+    for config in ConfigModel.query.all():
```
Comment (kubantjan) [inline txmatching/database/services/patient_service.py:425]: tady bych minimalne bral jen ty z relevantniho txm eventu.

File: txmatching/database/services/patient_service.py [hunk 5]
```diff
@@ -422,6 +422,18 @@ def delete_donor_recipient_pair(donor_id: int, txm_event_id: int):
...
+        if maybe_recipient is not None and maybe_recipient.db_id in config.parameters["required_patient_db_ids"]:
```
Comment (kubantjan) [inline txmatching/database/services/patient_service.py:426]: proc tady nekontrolujes i jestli to id neni mezi manual recipient donor scores?

File: txmatching/solve_service/solve_from_configuration.py [hunk 6]
```diff
@@ -20,6 +20,16 @@ def solve_from_configuration(config_parameters: ConfigParameters, txm_event: Txm
 
 
 def _solve_from_configuration_unsafe(config_parameters: ConfigParameters, txm_event: TxmEvent) -> PairingResult:
+    # check if there are invalid patients in the config
+    for patient_db_id in config_parameters.required_patient_db_ids:
```
Comment (kubantjan) [inline txmatching/solve_service/solve_from_configuration.py:24]: tady bych spis udelal list comprehension nez pouzival remove, to je docela draha operace, tady ji vyuzivas podle me zbytecne.

Comment (kubantjan) [reply txmatching/solve_service/solve_from_configuration.py:24]: neco jako config_parameters.required_patient_db_ids = [db_id for db_id in config_parameters.required_patient_db_ids if db_id in active_and_valid_recipients_dict]

Comment (kubantjan) [reply txmatching/solve_service/solve_from_configuration.py:24]: a podobne o radek niz

File: tests/web/test_patient_api.py [hunk 7]
```diff
@@ -773,7 +773,15 @@ def test_default_config_not_working_when_patients_change(self):
         txm_event_db_id = self.fill_db_with_patients()
 
         # 2. add configuration with required patients
-        configuration = ConfigParameters(required_patient_db_ids=[1, 2])
+        configuration = ConfigParameters(
```
Comment (krllstdn) [inline tests/web/test_patient_api.py:776]: Zde jsem si vsimnul ze my filtrujeme nevalidni pacienty z `manual_donor_recipient_score`, ale netestujeme zda to funguje nebo ne.

**Review by kubantjan — APPROVED** [review]

**Pull request merged**